pytorch_lightning/loops/epoch/evaluation_epoch_loop.py

```diff
@@ -134,10 +133,13 @@ def advance(
         self.trainer.logger_connector.update_eval_step_metrics()
 
         # track epoch level outputs
-        if self._should_track_batch_outputs_for_epoch_end():
-            output = recursive_detach(output, to_cpu=self.trainer.move_metrics_to_cpu)
-            if output is not None:
-                self.outputs.append(output)
+        if self._should_track_batch_outputs_for_epoch_end() and output is not None:
+            self.outputs.append(output)
+
+        if self.trainer.move_metrics_to_cpu:
+            # the evaluation step output is not moved as they are not considered "metrics"
+            assert self.trainer._results is not None
```

Comment:
why do we assert this? If somebody doesn't log metrics during validation but during training only and still uses the flag, it would error out here. Why would this be expected?

Reply:
I think it's due to mypy.

Reply:
Exactly. `Trainer._results` is `Optional` because the predict loop does not define it, but all other loops do

Reply:
wouldn't

```
if self.trainer._results is not None:
    self.trainer._results.cpu()
```

work?
Seems slightly cleaner than asserting, but as long as we are sure this is never triggered, I'm also fine with the current way

Reply:
I'd prefer that if that was something that could happen, but it's not.

I use assertions because they are short, the other option would be.

```python
if self.trainer.move_metrics_to_cpu:
    if self.trainer._results is None:
        raise InternalError("This shouldn't have happened")
    self.trainer._results.cpu()
```